core/src/state/world_state.rs

```diff
@@ -65,31 +65,73 @@ impl StateHandle {
             .write()
             .expect("Could not acquire world_state lock.")
     }
+}
 
-    /** Block asynchronously until the given sequence number is reached. */
-    pub async fn wait_for_seq(&self, sequence: u64) {
-        if self.state().logical_time >= sequence {
-            return;
-        }
-        let receiver = { self.write_state().get_listener(sequence) };
-        receiver.notified().await;
+/*
+We want a thing that waits until notified, and importantly waits
+for that notify on all of its replicas.
+The default notifier from tokio does not behave the way we need
+since, if multiple clones are made, notify_one only causes the first
+to return immediately when it invokes the notification and awaits.
+notify_waiters does not cause even the first one to do that.
+Therefore, we implement our own notifier that can be closed and
+will return immediately on all its dependent notifies.
+*/
+#[derive(Debug, Clone)]
+pub struct ClosableNotifier {
+    notifier: Sender<()>,
+}
+
+#[derive(Debug)]
+pub struct ClosableNotify(Receiver<()>);
+
+impl ClosableNotify {
+    pub async fn notified(&mut self) {
+        let _ = self.0.recv().await;
+    }
+}
+
+impl ClosableNotifier {
+    fn new() -> Self {
+        let (tx, _rx) = tokio::sync::broadcast::channel(128);
+        Self { notifier: tx }
+    }
+
+    fn notify(&self) -> ClosableNotify {
+        ClosableNotify(self.notifier.subscribe())
+    }
+
+    fn notify_waiters(&mut self) {
+        let _ = self.notifier.send(());
     }
 }
 
 #[derive(Default, Debug)]
 pub struct WorldState {
     logical_time: u64,
     pub clusters: BTreeMap<ClusterName, ClusterState>,
-    listeners: HashMap<u64, Arc<Notify>>,
+    listeners: RwLock<HashMap<u64, ClosableNotifier>>,
```

Comment:
let's just use a `DashMap` instead of an `RwLock<HashMap<..>>`

Reply:
I was running into a lot of deadlocks with dashmap, RwLock<HashMap> is much easier to understand.